Get single parameter

rohbotics opened the pull request:
Added an API for getting a single parameter.

Note: this PR is based on master, and might conflict with my other PR

wjwwood commented:
Should this function have an optional "default value" parameter? Or should we have a separate function that provides that syntax sugar?

wjwwood commented:
Sorry, looks like merging the other pr makes this one conflict. You'll need to rebase it.

rohbotics commented:
@wjwwood Rebased

Do you think the default value should be on all three versions of the function, or just on the one that returns a parameter?

dirk-thomas commented:
Adding more functions (in this case three different signatures) might be helpful for the user but imo it increases the long term maintenance burden since actual functionality and syntactic sugar is mixed (as I mentioned in https://github.com/ros2/rclcpp/pull/233#issuecomment-230037130).

From my point of view we should have one component with a minimal interface providing and implementing all the core functionality (call it a "micro kernel"). Any additional API for syntactic sugar should be separated.

rohbotics commented:
I thought that the minimal interface as supposed to be provided by rcl, and that rclcpp is the more user-friendly frontend.

If we want to keep the node class minimal, we could move all parameter functionality to the `parameter_service`, and have it have all the syntactic sugar.

That way any node that doesn't pull in the `parameter_service` isn't a parameter server, so doesn't have any internal map for storing parameters, which is currently in the node. We can also move the `parameter_events` publisher to the `parameter_service`, (Connects to #54). 

The downside to this approach would be that in order to pass a handle around for creating publishers/subscribers and getting/setting parameters, you would need a pointer to both the `parameter_service` and the node. We could mitigate this by having the node store a smart_ptr to a passed in `parameter_service`, or have a function that makes a `parameter_service` internal to the node.

rohbotics commented:
Would you say that maintainability would be increased if instead of the 3 signatures being completely independent/freestanding, they were composed on top of each-other?

rohbotics commented:
Sorry, I hit the close and comment instead of the comment button

wjwwood commented:
@rohbotics I have a different opinion from @dirk-thomas: this is the highest level interface and the appropriate place to have convenience functions. We can choose to back these functions with a minimal interface in the future if we choose. I do agree that you should try to have the functions call each other where possible. It looks like at least two of them are basically duplicates.

rohbotics commented:
How should I handle the locking between functions if they call each other?
Right now they all lock the same mutex, so if one of them locks it and then calls another, the callee will wait forever on the mutex, as the caller is holding it until it returns.

tfoote commented:
The higher level one does not need to lock if it's just calling the other one that locks.

rohbotics commented:
I made all the functions lock, then call a common non-locking private function.

wjwwood commented on rclcpp/include/rclcpp/node_impl.hpp:
```diff
@@ -368,6 +368,18 @@ void Node::register_param_change_callback(CallbackT && callback)
+
+  rclcpp::parameter::ParameterVariant parameter_variant(parameter);
+  bool result = get_parameter_(name, parameter_variant);
+  parameter = parameter_variant.get_value<ParameterT>();
```
If `result` is `false` will this not raise?

rohbotics replied:
This should not throw if `result == false`
That enables use like this
```
double foo = 0.1;
if (!get_parameter("foo", foo) {
  set_parameter("foo", foo);
}
```
We do throw if get_value fails because of differing types.

wjwwood replied:
But if `result` is `false`, then the `ParameterVariant` will be "not set" right? So `get_value` will raise I think (that was what I was referring to before).

wjwwood replied:
After talking with @rohbotics I think I might be wrong about `get_value` throwing because the `ParameterVariant` is seeded with the given `parameter` value.

rohbotics replied:
The seeding doesn't compile with anything other than a string, so looks like I need to rethink this a little bit.

wjwwood commented on rclcpp/include/rclcpp/node_impl.hpp:
```diff
@@ -368,6 +368,18 @@ void Node::register_param_change_callback(CallbackT && callback)
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  rclcpp::parameter::ParameterVariant parameter_variant(parameter);
+  bool result = get_parameter_(name, parameter_variant);
```
Could you not call the `bool get_parameter(std::string, ParameterVariant)` version of the function here and avoid the mutex lock two lines above?

wjwwood commented on rclcpp/src/rclcpp/node.cpp:
```diff
@@ -242,6 +242,39 @@ Node::get_parameters(
+}
+
+bool Node::get_parameter_(const std::string & name,
+  rclcpp::parameter::ParameterVariant & parameter) const
```
Hmm, maybe you could expand on why you don't just lock this function as @tfoote suggested? Seems to me that you could move the lock in the other three functions to this one without any issues.

wjwwood commented:
Also, I'd like to clarify that I agree with @dirk-thomas in that we need to restructure the Node class and make a separation between high level interfaces and the foundation of those interfaces. However, I see it as a larger task that goes beyond these couple of new functions, and that it shouldn't be a pre-requisite for opportunistically taking some contributions like this one.

rohbotics commented:
@wjwwood I cleaned up a little, and we don't need the non-locking private function

I added tests here https://github.com/ros2/system_tests/pull/158 as we discussed

wjwwood commented on rclcpp/src/rclcpp/node.cpp:
```diff
@@ -242,6 +242,31 @@ Node::get_parameters(
+  if (get_parameter(name, parameter)) {
+    return parameter;
+  } else {
+    throw std::out_of_range("Parameter: " + name + " does not exist in map");
```
nitpick: would you change this to `"Parameter '" + name + "' not set"`? We usually wrap variables output into string inside of quotes in case `name` is an empty string, for example. Also, the fact that a map is used is an implementation detail.

wjwwood commented:
Other than my comment, this lgtm. I still need to review the tests you added (thanks for those btw), and then run CI.

rohbotics commented:
I fixed the exception message

wjwwood commented on rclcpp/src/rclcpp/node.cpp:
```diff
@@ -242,6 +242,31 @@ Node::get_parameters(
+  if (get_parameter(name, parameter)) {
+    return parameter;
+  } else {
+    throw std::out_of_range("Parameter '" + name + "' not set");
```
Also, `out_of_range` may not be appropriate here. Instead maybe a custom exception should be used... I think if this were a method on a "ParameterSet" or "ParameterMap" object it might make sense because `out_of_range` is the functional reason the parameter cannot be returned (the given key resulted in a "key error"), but it's not really the non-functional reason. In the context of a user asking for a parameter, it might make sense to throw a more domain specific exception type.

I've not convinced myself enough to suggest that you should change it right now, but maybe others can give their opinions.

wjwwood commented:
I started CI for you:

- Linux:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=1678)](http://ci.ros2.org/job/ci_linux/1678/)
- OS X:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=1248)](http://ci.ros2.org/job/ci_osx/1248/)

wjwwood commented:
@rohbotics looks like your new tests have some compiler warnings: http://ci.ros2.org/job/ci_linux/1678/warnings20Result/new/

wjwwood commented:
I kicked off a new Linux CI to check the warnings:

- Linux:
  - [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=1679)](http://ci.ros2.org/job/ci_linux/1679/)

dirk-thomas commented:
One new warning on Windows: http://ci.ros2.org/job/ci_windows/1609/warnings39Result/new/

rohbotics commented:
Fixed warning and test on Windows

@wjwwood, looks like on windows you have to use C++11's <cstdint> instead of C99's <stdint.h>, or you need a specific include order

wjwwood commented:
@rohbotics Yeah, makes sense. CI looks good, thanks for following up on this.
